File: src/lib/canvasAssetRefs.ts
import type { FlowNodeData } from "@/lib/types";
import type { Node } from "@xyflow/react";

function maybeInsertAssetPath(out: Set<string>, raw: string) {
  const t = raw.trim().replace(/\\/g, "/");
  if (t.startsWith("assets/") && !t.includes("..")) {
    out.add(t);
  }
}

function walkValue(value: unknown, out: Set<string>) {
  if (typeof value === "string") {
    maybeInsertAssetPath(out, value);
    return;
  }
  if (Array.isArray(value)) {
    for (const item of value) walkValue(item, out);
    return;
  }
  if (value && typeof value === "object") {
    for (const v of Object.values(value as Record<string, unknown>)) {
      walkValue(v, out);
    }
  }
}

/** 从节点 data（含 scriptBeats / timelineClips 等嵌套字段）收集 assets 相对路径 */
export function collectAssetRelPathsFromNodeData(data: FlowNodeData | undefined): string[] {
  const out = new Set<string>();
  walkValue(data, out);
  return [...out];
}

export function collectAssetRelPathsFromNodes(nodes: Node<FlowNodeData>[]): string[] {
  const out = new Set<string>();
  for (const node of nodes) {
    for (const p of collectAssetRelPathsFromNodeData(node.data)) {
      out.add(p);
    }
  }
  return [...out];
}

File: src-tauri/src/canvas_asset_refs.rs
//! 从画布 JSON 收集工程内素材相对路径（用于删除节点后 GC）

use serde_json::Value;
use std::collections::HashSet;

fn maybe_insert_asset_path(out: &mut HashSet<String>, raw: &str) {
    let t = raw.trim().replace('\\', "/");
    if t.starts_with("assets/") && !t.contains("..") {
        out.insert(t);
    }
}

fn walk_value(value: &Value, out: &mut HashSet<String>) {
    match value {
        Value::String(s) => maybe_insert_asset_path(out, s),
        Value::Array(arr) => {
            for v in arr {
                walk_value(v, out);
            }
        }
        Value::Object(map) => {
            for v in map.values() {
                walk_value(v, out);
            }
        }
        _ => {}
    }
}

/// 递归扫描 JSON 中所有 `assets/...` 字符串引用
pub fn collect_asset_rel_paths(value: &Value) -> HashSet<String> {
    let mut out = HashSet::new();
    walk_value(value, &mut out);
    out
}

/// 从画布 nodes 数组收集仍被引用的素材路径
pub fn collect_asset_rel_paths_from_nodes(nodes: &Value) -> HashSet<String> {
    collect_asset_rel_paths(nodes)
}
